WorkflowInstanceTokenTest#shouldCompleteInstanceAfterInterruptingBoundaryEventTriggered is flaky

```
15:53:36 [1 - Java Tests] [ERROR] Tests run: 14, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 6.906 s <<< FAILURE! - in io.zeebe.broker.workflow.WorkflowInstanceTokenTest
15:53:36 [1 - Java Tests] [ERROR] shouldCompleteInstanceAfterInterruptingBoundaryEventTriggered(io.zeebe.broker.workflow.WorkflowInstanceTokenTest)  Time elapsed: 5.041 s  <<< ERROR!
15:53:36 [1 - Java Tests] io.zeebe.test.util.stream.StreamWrapperException: No event found matching the criteria
15:53:36 [1 - Java Tests] 	at java.util.Optional.orElseThrow(Optional.java:290)
15:53:36 [1 - Java Tests] 	at io.zeebe.test.util.stream.StreamWrapper.getFirst(StreamWrapper.java:80)
15:53:36 [1 - Java Tests] 	at io.zeebe.broker.workflow.WorkflowInstanceTokenTest.assertThatWorkflowInstanceCompletedAfter(WorkflowInstanceTokenTest.java:455)
15:53:36 [1 - Java Tests] 	at io.zeebe.broker.workflow.WorkflowInstanceTokenTest.shouldCompleteInstanceAfterInterruptingBoundaryEventTriggered(WorkflowInstanceTokenTest.java:308)
15:53:36 [1 - Java Tests] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
15:53:36 [1 - Java Tests] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
15:53:36 [1 - Java Tests] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
15:53:36 [1 - Java Tests] 	at java.lang.reflect.Method.invoke(Method.java:498)
15:53:36 [1 - Java Tests] 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
15:53:36 [1 - Java Tests] 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
15:53:36 [1 - Java Tests] 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
15:53:36 [1 - Java Tests] 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
15:53:36 [1 - Java Tests] 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
15:53:36 [1 - Java Tests] 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
15:53:36 [1 - Java Tests] 	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
15:53:36 [1 - Java Tests] 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
15:53:36 [1 - Java Tests] 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
15:53:36 [1 - Java Tests] 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
15:53:36 [1 - Java Tests] 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
15:53:36 [1 - Java Tests] 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
15:53:36 [1 - Java Tests] 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
15:53:36 [1 - Java Tests] 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
15:53:36 [1 - Java Tests] 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
15:53:36 [1 - Java Tests] 	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
15:53:36 [1 - Java Tests] 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
15:53:36 [1 - Java Tests] 	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
15:53:36 [1 - Java Tests] 	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
15:53:36 [1 - Java Tests] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
15:53:36 [1 - Java Tests] 	at org.junit.runners.Suite.runChild(Suite.java:128)
15:53:36 [1 - Java Tests] 	at org.junit.runners.Suite.runChild(Suite.java:27)
15:53:36 [1 - Java Tests] 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
15:53:36 [1 - Java Tests] 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
15:53:36 [1 - Java Tests] 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
15:53:36 [1 - Java Tests] 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
15:53:36 [1 - Java Tests] 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
15:53:36 [1 - Java Tests] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
15:53:36 [1 - Java Tests] 	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
15:53:36 [1 - Java Tests] 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
15:53:36 [1 - Java Tests] 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeLazy(JUnitCoreWrapper.java:119)
15:53:36 [1 - Java Tests] 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:87)
15:53:36 [1 - Java Tests] 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
15:53:36 [1 - Java Tests] 	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)
15:53:36 [1 - Java Tests] 	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
15:53:36 [1 - Java Tests] 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
15:53:36 [1 - Java Tests] 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
15:53:36 [1 - Java Tests] 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
```

--- a/broker-core/src/test/java/io/zeebe/broker/workflow/WorkflowInstanceTokenTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/workflow/WorkflowInstanceTokenTest.java
@@ -26,6 +26,7 @@ import io.zeebe.exporter.api.record.value.IncidentRecordValue;
 import io.zeebe.exporter.api.record.value.WorkflowInstanceRecordValue;
 import io.zeebe.model.bpmn.Bpmn;
 import io.zeebe.protocol.intent.IncidentIntent;
+import io.zeebe.protocol.intent.JobIntent;
 import io.zeebe.protocol.intent.WorkflowInstanceIntent;
 import io.zeebe.test.broker.protocol.clientapi.ClientApiRule;
 import io.zeebe.test.broker.protocol.clientapi.PartitionTestClient;
@@ -43,10 +44,10 @@ import org.junit.rules.RuleChain;
 
 public class WorkflowInstanceTokenTest {
 
-  private static EmbeddedBrokerRule brokerRule = new EmbeddedBrokerRule();
-  private static ClientApiRule apiRule = new ClientApiRule(brokerRule::getClientAddress);
+  private static final EmbeddedBrokerRule BROKER_RULE = new EmbeddedBrokerRule();
+  private static final ClientApiRule API_RULE = new ClientApiRule(BROKER_RULE::getClientAddress);
 
-  @ClassRule public static RuleChain ruleChain = RuleChain.outerRule(brokerRule).around(apiRule);
+  @ClassRule public static RuleChain ruleChain = RuleChain.outerRule(BROKER_RULE).around(API_RULE);
 
   @Rule
   public RecordingExporterTestWatcher recordingExporterTestWatcher =
@@ -57,7 +58,8 @@ public class WorkflowInstanceTokenTest {
 
   @Before
   public void setUp() {
-    testClient = apiRule.partitionClient();
+    BROKER_RULE.getClock().reset();
+    testClient = API_RULE.partitionClient();
     processId = Strings.newRandomValidBpmnId();
   }
 
@@ -191,7 +193,7 @@ public class WorkflowInstanceTokenTest {
   @Test
   public void shouldCompleteInstanceAfterTimerIntermediateCatchEvent() {
     // given
-    brokerRule.getClock().pinCurrentTime();
+    BROKER_RULE.getClock().pinCurrentTime();
 
     testClient.deploy(
         Bpmn.createExecutableProcess(processId)
@@ -210,7 +212,7 @@ public class WorkflowInstanceTokenTest {
     // when
     testClient.completeJobOfType(workflowInstanceKey, "task");
 
-    brokerRule.getClock().addTime(Duration.ofSeconds(1));
+    BROKER_RULE.getClock().addTime(Duration.ofSeconds(1));
 
     // then
     assertThatWorkflowInstanceCompletedAfter(workflowInstanceKey, "end-2");
@@ -300,9 +302,12 @@ public class WorkflowInstanceTokenTest {
         testClient.createWorkflowInstance(r -> r.setBpmnProcessId(processId)).getInstanceKey();
 
     // when
-    testClient.receiveElementInState(
-        workflowInstanceKey, "task", WorkflowInstanceIntent.ELEMENT_ACTIVATED);
-    brokerRule.getClock().addTime(Duration.ofSeconds(1));
+    testClient
+        .receiveJobs()
+        .withWorkflowInstanceKey(workflowInstanceKey)
+        .withIntent(JobIntent.CREATED)
+        .getFirst();
+    BROKER_RULE.getClock().addTime(Duration.ofSeconds(1));
 
     // then
     assertThatWorkflowInstanceCompletedAfter(workflowInstanceKey, "end-2");
@@ -326,9 +331,12 @@ public class WorkflowInstanceTokenTest {
         testClient.createWorkflowInstance(r -> r.setBpmnProcessId(processId)).getInstanceKey();
 
     // when
-    testClient.receiveElementInState(
-        workflowInstanceKey, "task-1", WorkflowInstanceIntent.ELEMENT_ACTIVATED);
-    brokerRule.getClock().addTime(Duration.ofSeconds(1));
+    testClient
+        .receiveJobs()
+        .withWorkflowInstanceKey(workflowInstanceKey)
+        .withIntent(JobIntent.CREATED)
+        .getFirst();
+    BROKER_RULE.getClock().addTime(Duration.ofSeconds(1));
     testClient.completeJobOfType(workflowInstanceKey, "task-2");
     testClient.completeJobOfType(workflowInstanceKey, "task-1");
 

--- a/test-util/src/main/java/io/zeebe/test/util/record/RecordingExporter.java
+++ b/test-util/src/main/java/io/zeebe/test/util/record/RecordingExporter.java
@@ -24,7 +24,6 @@ import io.zeebe.exporter.api.record.value.JobRecordValue;
 import io.zeebe.exporter.api.record.value.MessageRecordValue;
 import io.zeebe.exporter.api.record.value.MessageStartEventSubscriptionRecordValue;
 import io.zeebe.exporter.api.record.value.MessageSubscriptionRecordValue;
-import io.zeebe.exporter.api.record.value.RaftRecordValue;
 import io.zeebe.exporter.api.record.value.TimerRecordValue;
 import io.zeebe.exporter.api.record.value.VariableDocumentRecordValue;
 import io.zeebe.exporter.api.record.value.VariableRecordValue;
@@ -40,7 +39,6 @@ import io.zeebe.protocol.intent.JobIntent;
 import io.zeebe.protocol.intent.MessageIntent;
 import io.zeebe.protocol.intent.MessageStartEventSubscriptionIntent;
 import io.zeebe.protocol.intent.MessageSubscriptionIntent;
-import io.zeebe.protocol.intent.RaftIntent;
 import io.zeebe.protocol.intent.TimerIntent;
 import io.zeebe.protocol.intent.VariableDocumentIntent;
 import io.zeebe.protocol.intent.VariableIntent;
@@ -52,6 +50,10 @@ import java.util.List;
 import java.util.Spliterator;
 import java.util.Spliterators;
 import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 
@@ -59,15 +61,17 @@ public class RecordingExporter implements Exporter {
 
   private static final List<Record<?>> RECORDS = new CopyOnWriteArrayList<>();
 
-  private static final Object EXPORT_MONITOR = new Object();
+  private static final Lock LOCK = new ReentrantLock();
+  private static final Condition IS_EMPTY = LOCK.newCondition();
+
   private static final long MAX_WAIT = Duration.ofSeconds(5).toMillis();
 
   @Override
   public void export(final Record record) {
+    LOCK.lock();
     RECORDS.add(record);
-    synchronized (EXPORT_MONITOR) {
-      EXPORT_MONITOR.notifyAll();
-    }
+    IS_EMPTY.signal();
+    LOCK.unlock();
   }
 
   public static List<Record<?>> getRecords() {
@@ -75,7 +79,9 @@ public class RecordingExporter implements Exporter {
   }
 
   public static void reset() {
+    LOCK.lock();
     RECORDS.clear();
+    LOCK.unlock();
   }
 
   @SuppressWarnings("unchecked")
@@ -93,14 +99,6 @@ public class RecordingExporter implements Exporter {
         Spliterators.spliteratorUnknownSize(new RecordIterator(), Spliterator.ORDERED);
     return new RecordStream(
         StreamSupport.stream(spliterator, false).map(r -> (Record<RecordValue>) r));
-  }
-
-  public static RaftRecordStream raftRecords() {
-    return new RaftRecordStream(records(ValueType.RAFT, RaftRecordValue.class));
-  }
-
-  public static RaftRecordStream raftRecords(final RaftIntent intent) {
-    return raftRecords().withIntent(intent);
   }
 
   public static MessageSubscriptionRecordStream messageSubscriptionRecords() {
@@ -228,18 +226,22 @@ public class RecordingExporter implements Exporter {
 
     @Override
     public boolean hasNext() {
-      if (isEmpty()) {
-        // block haven't got enough records yet
-        try {
-          synchronized (EXPORT_MONITOR) {
-            EXPORT_MONITOR.wait(MAX_WAIT);
+      LOCK.lock();
+      try {
+        long now = System.currentTimeMillis();
+        final long endTime = now + MAX_WAIT;
+        while (isEmpty() && endTime > now) {
+          final long waitTime = endTime - now;
+          try {
+            IS_EMPTY.await(waitTime, TimeUnit.MILLISECONDS);
+          } catch (final InterruptedException e) {
           }
-        } catch (final InterruptedException e) {
-          throw new RuntimeException(e);
+          now = System.currentTimeMillis();
         }
+        return !isEmpty();
+      } finally {
+        LOCK.unlock();
       }
-
-      return !isEmpty();
     }
 
     @Override

